services/controlplane/features/client.go

```diff
@@ -0,0 +1,147 @@
+package features
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"runtime"
+	"sort"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+)
+
+var (
+	DefaultTimeout = 30 * time.Second
+	MaxRetries     = uint64(3)
+)
+
+type OptFn func(c *Client)
+
+func WithURL(url string) OptFn {
+	return func(c *Client) {
+		c.url = url
+	}
+}
+
+func WithHTTPClient(httpClient *http.Client) OptFn {
+	return func(c *Client) {
+		c.client = httpClient
+	}
+}
+
+func WithTimeout(timeout time.Duration) OptFn {
+	return func(c *Client) {
+		c.client.Timeout = timeout
+	}
+}
+
+type Client struct {
+	client   *http.Client
+	ua       string
+	url      string
+	identity identity.Identifier
+}
+
+type payload struct {
+	Components []component `json:"components"`
+}
+
+type component struct {
+	Name     string   `json:"name"`
+	Features []string `json:"features"`
+}
+
+func hostname() string {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "unknown"
+	}
+	return hostname
+}
+
+func New(identity identity.Identifier, fns ...OptFn) *Client {
+	c := &Client{
+		client: &http.Client{
+			Timeout: DefaultTimeout,
+		},
+		url:      config.GetEnv("CONFIG_BACKEND_URL", "https://api.rudderlabs.com"),
+		identity: identity,
+		ua:       fmt.Sprintf("Go-http-client/1.1; %s; control-plane/features; %s", runtime.Version(), hostname()),
+	}
+
+	for _, fn := range fns {
+		fn(c)
+	}
+
+	return c
+}
+
+type PerComponent map[string][]string
+
+func (c *Client) Send(ctx context.Context, components PerComponent) error {
```

Comment:
Not sure it's worth having a type for `components` here. Not strong on it but consider that we won't be able to decouple much when using this stuff (e.g. possibility to create local interface without depending on this package):

```go
type client interface {
     Send(ctx context.Context, components features.PerComponent) error
}
```

:point_up: still depends on `features` versus...

```go
type client interface {
     Send(ctx context.Context, components map[string][]string) error
}
```